File: TensorFlowKeyboard/android/app/src/main/kotlin/com/tensorflowkeyboard/app/capture/InMemoryMessageBuffer.kt
package com.tensorflowkeyboard.app.capture

class InMemoryMessageBuffer(
    private val capacity: Int,
) {
    private val items = ArrayDeque<CaptureEvent>(capacity)

    @Synchronized
    fun add(event: CaptureEvent) {
        if (items.size == capacity) {
            items.removeFirst()
        }
        items.addLast(event)
    }

    @Synchronized
    fun snapshot(): List<CaptureEvent> {
        return items.toList().asReversed()
    }
}

File: TensorFlowKeyboard/ios/KeyboardExtension/InMemoryMessageBuffer.swift
import Foundation

final class InMemoryMessageBuffer {
    private let capacity: Int
    private var items: [CaptureEvent] = []

    init(capacity: Int = 20) {
        self.capacity = capacity
    }

    func add(_ event: CaptureEvent) {
        if items.count == capacity {
            items.removeFirst()
        }
        items.append(event)
    }

    func snapshot() -> [CaptureEvent] {
        return items.reversed()
    }
}
